flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/RestartPipelinedRegionStrategy.java

```diff
@@ -165,6 +181,33 @@ private void buildOneRegionForAllVertices() {
 		for (FailoverVertex vertex : topology.getFailoverVertices()) {
 			vertexToRegionMap.put(vertex.getExecutionVertexID(), region);
 		}
+
+		buildRegionInputsAndOutputs();
+	}
+
+	private void buildRegionInputsAndOutputs() {
+		for (FailoverRegion region : regions.keySet()) {
+			IdentityHashMap<FailoverRegion, Object> consumers = new IdentityHashMap<>();
+			Set<IntermediateResultPartitionID> inputs = new HashSet<>();
+			Set<ExecutionVertexID> consumerVertices = new HashSet<>();
+			Set<FailoverVertex> regionVertices = region.getAllExecutionVertices();
+			regionVertices.forEach(v -> {
```

Comment:
related: https://github.com/apache/flink/pull/8430#discussion_r284194311

The idea here is fine. I'm wondering if we could merge this logic into the initial region building; we already iterate over all inputs there anyway. I think we can even detect region boundaries relatively easy (if an input edge is not pipelined you have a boundary, and you even know the (temporary) regions)

Reply:
I tried the merged logic and find the region building becomes even slower.
I think this is due to the merged logic saved some iteration overhead, but added some more work, including building some legacy relations and later deprecating them in region merging.